feat(core): add migration for invalid two-way bindings

crisbeto opened the pull request:
As a part of #54154, an old parser behavior came up where two-way bindings were parsed by appending `= $event` to the event side. This was problematic, because it allowed some non-writable expressions to be passed into two-way bindings.

These changes introduce a migration that will change the two-way bindings into two separate input/output bindings that represent the old behavior so that in a future version we can throw a parser error for the invalid expressions.

```ts
// Before
@Component({
  template: `<input [(ngModel)]="a && b"/>`
})
export class MyComp {}

// After
@Component({
  template: `<input [ngModel]="a && b" (ngModelChange)="a && (b = $event)"/>`
})
export class MyComp {}
```

devversion commented on packages/core/schematics/test/invalid_two_way_bindings_spec.ts:
```diff
@@ -0,0 +1,341 @@
+           'template: `<input [(ngModel)]="a || b" (ngModelChange)="foo($event)"/>`');
+     });
+
+  it('should not migrate an invalid expression if a property binding for the same event exists',
```
Nit, but honestly neither is making me feel good 😄 

```suggestion
  it('should not migrate an invalid expression if a one-way property binding for the same binding exists',
```

devversion reviewed (approved):
Nice— and wow. The behavior is pretty magical

crisbeto committed:
feat(core): add migration for invalid two-way bindings

As a part of #54154, an old parser behavior came up where two-way bindings were parsed by appending `= $event` to the event side. This was problematic, because it allowed some non-writable expressions to be passed into two-way bindings.

These changes introduce a migration that will change the two-way bindings into two separate input/output bindings that represent the old behavior so that in a future version we can throw a parser error for the invalid expressions.

```ts
// Before
@Component({
  template: `<input [(ngModel)]="a && b"/>`
})
export class MyComp {}

// After
@Component({
  template: `<input [ngModel]="a && b" (ngModelChange)="a && (b = $event)"/>`
})
export class MyComp {}
```

crisbeto commented:
Caretaker note: CI failures are unrelated and present in `main`.

pkozlowski-opensource commented:
This PR was merged into the repository by commit fb540e169a78a61f38d611f538eea8fdb0971f1d.
